File: ho-woo/object-orientated/euler_16_power_digit_sum.java/euler_16_power_digit_sum.java
/* Digit represents a digit */
class Digit {
    int value;              // Value of this Digit node
    Digit prev = null;      // Previous node in the linked list
    Digit next = null;      // Next node in the linked list

    Digit(int value) {
        this.value = value;
    }

    public String toString() {
        return Integer.toString(this.value);
    }
}

/* DigitList is a linked list of Digits representing the digits in the power of 2.
   The numbers are stored in a reverse order, where the last digit of the number is at the head.
*/
class DigitList {
    Digit head;     // First node of the linked list
    Digit tail;     // Last node of the linked list 

    DigitList(int head) {
        this.head = new Digit(head);
        this.tail = this.head;
    }

    // Display the Digits in the linked list
    public String toString() {
        String output = "";

        Digit current = this.tail;
        do
        {
            output += current.value;
            current = current.prev;
        } while (current != null);
        return output;
    }
}

/* PowerDigitSum represents the sum of the digits of 2 power a given exponent */
class PowerDigitSum {
    final int base = 2;
    int exponent;
    DigitList digitList;
    int sum;

    // Constructor initializes the variables and call digitSum() to perform calculation
    PowerDigitSum(int exponent) {
        this.exponent = exponent;
        this.digitList = new DigitList(1);
        
        digitSum();     // Call digitSum() to calculate the power digit sum
    }

    // Calculate the power digit sum with the given exponent
    void digitSum() {
        powerToDigits();    // Call powerToDigits() to calculate the power and convert it to a DigitList

        // Calculate the sum of all Digits in the DigitList
        Digit current = this.digitList.head;
        do
        {
            this.sum += current.value;
            current = current.next;
        } while (current != null);
    }

    // Calculate the nth power of 2 where n is the given exponent and return a DigitList representing the power
    void powerToDigits() {
        int carry;
        Digit current, temp;

        /* Multiply the DigitList by 2 in each loop*/
        for (int i = 0; i < this.exponent; i++)
        {
            current = this.digitList.head;
            carry = 0;

            // Multiply the DigitList by 2
            while (current != null)
            {
                current.value = current.value * this.base + carry;
    
                if (current.value >= 10) 
                {
                    carry = current.value / 10;
                    current.value %= 10;
                }
                else
                {
                    carry = 0;
                }
                current = current.next;
            }

            // Insert carry at the tail
            if (carry > 0)
            {
                temp = new Digit(carry);
                temp.prev = this.digitList.tail;
                this.digitList.tail.next = temp;
                this.digitList.tail = temp;
            }
        }
    }
}

public class euler_16_power_digit_sum {
    public static void main(String[] args)
    {
        int exponent = 1000;    // Default exponent is 1000
        if (args.length > 0)
        {
            exponent = Integer.parseInt(args[0]);   // Get exponent from command line
        }

        PowerDigitSum powerDigitSum = new PowerDigitSum(exponent);  // Initialize a PowerDigitSum object with given exponent
        System.out.println(powerDigitSum.sum);  // Print result
    }
}

File: ho-woo/procedural/euler-16-power-digit-sum.c/euler-16-power-digit-sum.c
#include <stdio.h>
#include <stdlib.h>

/* Global variables */
const int base = 2;
int exponent;

/* Use a linked list of Digits to represent the large number (2 power exponent) */
/* struct Digit containing a digit and keep a record of the previous node and next node in the list */
typedef struct Digit
{
    int value;
    struct Digit* prev;
    struct Digit* next;
}Digit;

/* Calculate the sum of the digits of the number and print the sum */
void printSum(Digit* tail) {
    Digit *current = tail;  // Start from tail
    int sum = 0;
    /* Iterate through the digits and sum them */
    do
    {
        sum += current->value;  
        current = current->prev;
    } while (current != NULL);

    printf("%d\n", sum);
}

/* Calculate the number 2 powers the input exponent using a linked list of Digits.
   Return the tail pointer of the linked list of Digits representing the resulting number. */
Digit* powerToDigits(Digit* head, Digit* tail) {
    int carry;
    /* Allocate meomory for current and temp */
    Digit* current = malloc(sizeof(Digit));
    Digit* temp = malloc(sizeof(Digit));

    if (current == NULL || temp == NULL) {
        printf("Memory allocation failed.\n");
        exit(1);
    }

    /* Multiply the number in each loop. The number of iteration is the input exponent. */
    for (int i = 0; i < exponent; i++)
    {
        current = head;
        carry = 0;

        /* Multiply the number in each loop */
        while (current != NULL)
        {
            current->value = current->value * base + carry;   // Calculate the value in each digit

            /* Carrying */
            if (current->value >= 10) 
            {
                carry = current->value / 10;
                current->value %= 10;
            }
            else
            {
                carry = 0;
            }
            current = current->next;
        }

        /* Add carry to the tail if carry > 0 */
        if (carry > 0)
        {
            tail->next = malloc(sizeof(Digit));
            if (tail->next == NULL) {
                printf("Memory allocation failed.\n");
                exit(1);
            }

            tail->next->value = carry;
            tail->next->prev = tail;
            tail = tail->next;
        }
    }
    return tail;
}

int main(int argc, char **argv)
{
    exponent = 1000;    // Set default exponent to 1000 as required in the original question

    if (argc > 1)           // Take input from command line if a upper limit is given
    {
        exponent = atoi(argv[1]);
    }

    /* Initialize head pointer of the linked list */
    Digit *head = malloc(sizeof(Digit));
    if (head == NULL) {
        printf("Memory allocation failed.\n");
        return 1;
    }
    head->value = 1;
    head->prev = NULL;
    head->next = NULL;

    Digit *tail = head;     // Initialize tail pointer
    tail = powerToDigits(head, tail);   // Call powerToDigits() and receive the tail of the resulting number
    printSum(tail); // Call printSum() to calculate and display the sum of the digits in the number

    return 0;
}